### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by AgustinSilverfin**
First Draft of an authentication module. To use it:
- You will need to set first your firm id as an env var: `export SF_FIRM_ID=...`
- You can run the authorisation process: `.node_modules/sf-toolkit/bin/cli.js authorize`
- You should be able now to run any of the existing commands
 
Some comments:
- Credentials are now saved in a json file, still to decide where to store them.
- My idea is to make the firm id a required argument in the CLI, that way we could potentially handle different firms where the user has access to.
- Moved sf_api to a separate directory.

### CODE REVIEW

**Comment by AgustinSilverfin** [comment]
- Updated cli.js and how we handle CLI arguments.
- Now firm id is a required argument for most commands, and I think it should also be more clear and flexible now how the different commands work (cli.js)
- Also, commands names where changed. You check the new [Readme](https://github.com/silverfin/sf-toolkit/blob/auth/README.md) or use the built-in `--help` in the CLI

File: README.md [hunk 1]
```diff
@@ -27,82 +28,112 @@ Install `yarn` globally:
...
+
+```
+yarn init
+```
+
+Install `sf-toolkit` as a dependency of your project:
 
-You can either add the environment variables locally in the file so you only need to add them once, or define them directly in the terminal. 
+```
+yarn add https://github.com/silverfin/sf-toolkit.git
```
Comment (Oluwadamilareolusakin) [inline README.md:42]: We should eventually consider publishing an official npm package so they can add by doing something like `yarn add sf-toolkit` or `yarn add silverfin`

File: README.md [hunk 2]
```diff
@@ -27,82 +28,112 @@ Install `yarn` globally:
...
-touch .env
+  "scripts": {
+    "silverfin": "node ./node_modules/sf_toolkit/bin/cli.js"
+  },
 ```
 
-Add the following variables in  .env:
+## Add your API credentials as environmental variables
+
+You could use a new `.env` local file, or add them to an existing file in your system (e.g. `~/.bash-profile` or `~/.bashrc`)
```
Comment (Oluwadamilareolusakin) [inline README.md:57]: May want to include `.zshrc` in the list

File: README.md [hunk 3]
```diff
@@ -27,82 +28,112 @@ Install `yarn` globally:
...
+export SF_API_SECRET=...
 ```
 
-### Linux / Mac terminal
+If you are only using one firm, you could set it up as an environmental variable. That way, you won't need to pass the `--firm <firmid>` option every time you run a command.
 
 ```
-export SF_FIRM_ID=<firm_id>
-export SF_ACCESS_TOKEN=<access_token>
+export SF_FIRM_ID=...
```
Comment (Oluwadamilareolusakin) [inline README.md:67]: Okay for the first iteration, but you can also include a command in your package to set the firm/context they want to interact with, maybe `yarn silverfin -f firm_id_here` or `yarn silverfin -firm firm_id_here` which stores the current firm moving forward, it's more flexible than setting an env var or entering the firm id with every command

File: README.md [hunk 4]
```diff
@@ -27,82 +28,112 @@ Install `yarn` globally:
...
+
+## Import all existing reconciliations
 
 ```
-"scripts": {
-    "new-recon": "./node_modules/sf_toolkit/bin/cli.js new",
-    "import-recon": "./node_modules/sf_toolkit/bin/cli.js import",
-    "update-recon": "./node_modules/sf_toolkit/bin/cli.js persistReconciliationText"
-  }
+yarn silverfin import-all-reconciliations --firm <firm-id> --handle <handle>
```
Comment (Oluwadamilareolusakin) [inline README.md:99]: What's the difference between this and the import-reconciliation command, they look like they take the same set of arguments. Does the latter import the first reconciliation it finds in the firm with that handle and this one imports all reconciliations in a firm with a given handle?

**Review by Oluwadamilareolusakin — COMMENTED** [review]
The work done here is really great 👏  I'm curious about who the audience of this tool is and if what we should be building is SDKs instead of CLI tools, just curious about if this is meant to be used as a standalone tool and who is meant to be using it. If it's meant to be standalone, I'm curious as to why we are writing a yarn based tool as opposed to something more natural like `kubectl` or `gcloud` CLIs

File: README.md (hunk 1 above)
Comment (AgustinSilverfin) [reply README.md:42]: Yes, indeed. `yarn` or `npm run` should be temporary solutions until we publish an npm package

File: README.md (hunk 4 above)
Comment (AgustinSilverfin) [reply README.md:99]: Indeed, `import-all-reconciliations` is not developed yet, just assigned the command name as a placeholder. But it should import all the reconciliations existing in the firm. The intention is to be used when you are setting up your local repository, and you still have no reconciliations in it. Similar to `import-all-shared-parts`. But this shouldn't have the `--handle <handle>` argument, only `--firm <firmid>`

**Comment by Woetfin** [comment]
Great work! 🙌  The authentication module in the current way it works can be merged as far as I'm concerned. 

Some things we can (potentially) still improve upon in a later stage:
* Automatically receive the authorization code & authorized firm id back in the CLI
* Use the authorization url that already contains the `firm_id` to skip the firm selection step `https://live.getsilverfin.com/f/{{firm_id}}/oauth/authorize?client_id={{client_id}}&redirect_uri={{callback_url}}&response_type=code&scope={{scope}}`
* Publish as an npm package and make the `silverfin` command global
* Test integration in the CI workflow

**Review by Woetfin — APPROVED** [review]

**Review by dbobfin — APPROVED** [review]

**Review by TheSmartnik — APPROVED** [review]
Great job 👍  Everything looks find IMHO

**Pull request merged**